### ISSUE TITLE
ROCm: Flash/mem-efficient SDPA hangs during batched autoregressive generation with left-padded sequences

### ISSUE THREAD

**Issue by Finrandojin** [pytorch/pytorch#186174]
## Summary

Flash and memory-efficient SDPA backends hang indefinitely during batched autoregressive generation (`model.generate()`) when input sequences are left-padded to equal length. Disabling both backends and falling back to the math SDPA kernel resolves the hang. Single-sequence generation (no padding) works correctly with all backends.

## Environment

- **PyTorch:** 2.12.0+rocm7.2
- **ROCm:** 7.2.0 (HIP 7.2.53211)
- **GPU:** AMD Radeon RX 7900 XTX (gfx1100, 24 GB VRAM)
- **OS:** Ubuntu 24.04, kernel 6.8.0-117-generic
- **Python:** 3.10.18
- **transformers:** 4.57.3

## Reproduction

Reproducible with the [Qwen3-TTS](https://huggingface.co/Qwen/Qwen3-TTS-12Hz-1.7B-Base) model using the `qwen_tts` library (v0.1.1). The model uses SDPA attention (via HuggingFace transformers) and supports batched inference with left-padding.

```python
from qwen_tts import Qwen3TTSModel
import torchaudio

model = Qwen3TTSModel.from_pretrained(
    "Qwen/Qwen3-TTS-12Hz-1.7B-Base",
    device_map="cuda:0",
)

ref_audio, sr = torchaudio.load("reference.wav")  # any 16kHz mono WAV
ref_text = "Transcription of the reference audio."

# Single generation — works fine with all SDPA backends
audio_single = model.generate_voice_clone(
    text="This is a test sentence.",
    ref_audio=ref_audio,
    ref_text=ref_text,
)

# Batch generation — hangs indefinitely (left-pads shorter sequences)
audio_batch = model.generate_voice_clone(
    text=["Short text.", "This is a significantly longer sentence that will cause left-padding of the shorter text."],
    ref_audio=ref_audio,
    ref_text=ref_text,
)
```

### What the model does internally

When `text` is a list, [`Qwen3TTSModel.generate()`](https://github.com/QwenLM/Qwen3-TTS/blob/main/qwen_tts/core/models/modeling_qwen3_tts.py) builds per-sequence `talker_input_embeds`, then left-pads them to equal length with `torch.nn.utils.rnn.pad_sequence(..., padding_value=0.0)`. It creates a proper 2D attention mask (0 for padded positions, 1 for real) and passes both to `self.talker.generate()` — the standard HuggingFace `GenerationMixin.generate()` autoregressive loop.

The hang occurs inside this autoregressive loop. The model never emits EOS and generation runs indefinitely at high GPU utilization. There is no crash, no error, no timeout — it simply never returns.

## Workaround

Disabling flash and memory-efficient SDPA backends before batch generation forces the math backend, which handles the left-padded batched generation correctly:

```python
import torch

# Save original state
flash_was = torch.backends.cuda.flash_sdp_enabled()
mem_was = torch.backends.cuda.mem_efficient_sdp_enabled()

# Disable for batch generation
torch.backends.cuda.enable_flash_sdp(False)
torch.backends.cuda.enable_mem_efficient_sdp(False)

audio_batch = model.generate_voice_clone(
    text=["Short text.", "A much longer sentence requiring left-padding."],
    ref_audio=ref_audio,
    ref_text=ref_text,
)

# Restore
torch.backends.cuda.enable_flash_sdp(flash_was)
torch.backends.cuda.enable_mem_efficient_sdp(mem_was)
```

## Observations

- **Single-sequence generation** (no left-padding): Works with all SDPA backends.
- **Batched generation** (left-padded): Hangs with flash and/or mem-efficient SDPA enabled. Works with math-only SDPA.
- The hang is **deterministic** — same inputs always hang.
- The hang occurs regardless of batch size (tested with batch sizes 2, 5, 10, 50).
- The hang occurs regardless of text length — both short and long texts hang in batch mode.
- **NVIDIA users are not affected** — this appears specific to ROCm's SDPA kernel implementations.
- Previously worked with PyTorch + ROCm 6.3. The regression appeared after upgrading to ROCm 7.2 / PyTorch 2.12.0+rocm7.2.

## Notes

- A standalone minimal reproducer (without the model) using raw `F.scaled_dot_product_attention` with left-padded tensors and explicit masks does *not* reproduce the hang in isolation. The issue manifests specifically during the iterative KV-cache-based autoregressive generation loop, suggesting it may involve the interaction between the SDPA kernels and growing KV cache with padded positions over many steps.
- The model architecture uses GQA (16 Q-heads, 2 KV-heads, head_dim=64) which may be relevant to the kernel path selection.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @drisspg @liangel-02 @howardzhang-cv @include-yy

**Comment by naromero77amd** [comment]
Reproduced in the PyTorch nightly. @xinyazhang is investigating.

**Comment by xinyazhang** [comment]
The investigation may take some time since it requires additional debugging facilities established in either pytorch/aotriton.

**Comment by Finrandojin** [comment]
## Update: Root cause identified — not an SDPA bug

After further investigation, the batch generation hang reported here was **not caused by the SDPA kernels**. The actual root cause is the GPU's DPM (Dynamic Power Management) controller on RDNA3 (gfx1100) aggressively downclocking the shader engine between autoregressive generation steps under ROCm 7.x.

### What's actually happening

During `model.generate()`, each autoregressive step dispatches a short GPU kernel followed by a brief gap. The DPM controller interprets these gaps as idle time and drops the shader clock (sclk) from ~3000MHz to 40-100MHz within milliseconds. At these frequencies, the next kernel launch takes so long that it appears to hang — we observed 15+ minutes on sub-batches that normally complete in seconds.

This was confirmed by:
- `rocm-smi` showing sclk at 40-100MHz during active generation
- The same generation completing instantly when the GPU was at full clocks
- All software-level power controls (ppfeaturemask, DPM level locking, perf-level high/manual, ASPM, runtime PM) failing to prevent the downclocking

### Fix

Setting the GPU power profile to COMPUTE enforces a minimum clock frequency floor (1000MHz) that prevents the aggressive downclocking:

```bash
echo 5 | sudo tee /sys/class/drm/card1/device/pp_power_profile_mode
```

With the COMPUTE profile active, batch generation runs at full speed with stable clocks (~3000MHz).

### What this means for this issue

The SDPA backend disable workaround we described in the original report was coincidentally masking the problem — the math SDPA backend's lighter compute load made the clock drops less catastrophic but didn't actually fix anything. The real fix is the power profile change.

This appears to be a ROCm 7.x regression in how the DPM controller handles bursty autoregressive inference workloads on consumer RDNA GPUs. ROCm 6.x did not exhibit this behavior.

I'll leave this issue open in case the PyTorch team wants to investigate whether PyTorch could set the COMPUTE profile automatically when running on ROCm, but the issue is ultimately in the amdgpu driver's power management, not PyTorch's SDPA implementation.
